File: leetcode/course-schedule.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include "../utils.h"

using namespace std;

/**
https://leetcode.com/problems/course-schedule/

// Solution1. BFS Iteration (Topological Sort). Kahn's algorithm
// time: o(V+E). V(vertices) E(Edges)
// space: o(V+E). graph(V+E) + indegree(V)

1. build graph and count indegree
2. BSF to to order indepedent to depedent

// Solution2. DFS recursion (Topological Sort) with visited and instack to detact cycle
// time: o(V+E). V(vertices) E(edges). graph(V+E). cycle check(V+E)
// space: o(V+E). graph(V+E) recursion(V+E)

1. save as directed graph
2. DFS recursion call on each node. add visited to skip. push and pop instack to detact cycle

*/

class Solution {
public:
    bool canFinish_BFS_Iteration(int numCourses, vector<vector<int>>& prerequisites) {
        // build graph & count indegree
        vector<unordered_set<int>> graph(numCourses);
        vector<int> indegree(numCourses, 0);
        for (auto &preq:prerequisites) {
            graph[preq[1]].insert(preq[0]);
            indegree[preq[0]]++;
        }
        
        vector<int> bfs;
        for (int i=0; i<numCourses; ++i) {
            if (indegree[i] == 0) bfs.push_back(i);
        }
        
        for (int i=0; i<bfs.size(); ++i) {
            for (auto &adj:graph[bfs[i]]) {
                if (--indegree[adj] == 0) bfs.push_back(adj);
            }
        }
        
        return bfs.size() == numCourses;
    }
    
    bool hasCycle(vector<unordered_set<int>>& graph, unordered_set<int>& visited, unordered_set<int>& instack, int node) {
        if (instack.find(node) != instack.end()) return true;
        if (visited.find(node) != visited.end()) return false;
        
        visited.insert(node);
        instack.insert(node);
        for (auto &adj:graph[node]) {
            if (hasCycle(graph, visited, instack, adj)) return true;
        }
        
        instack.erase(node);
        return false;
    }
    
    bool canFinish_DSF_recursion(int numCourses, vector<vector<int>>& prerequisites) {
        vector<unordered_set<int>> graph(numCourses);
        for (auto &preq:prerequisites) {
            graph[preq[1]].insert(preq[0]);
        }
        
        unordered_set<int> visited;
        unordered_set<int> instack;
        for (int i=0;i<numCourses;++i) {
            if (hasCycle(graph, visited, instack, i)) return false;
        }
        
        return true;
    }
    
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        return canFinish_BFS_Iteration(numCourses, prerequisites);
        //return canFinish_DSF_recursion(numCourses, prerequisites);
    }
};

int main()
{
    auto input1 = vector<vector<int>> {{1,0}};
    assert(Solution().canFinish(2, input1) == (true));
    auto input2 = vector<vector<int>> {{1,0},{0,1}};
    assert(Solution().canFinish(2, input2) == (false));
    return 0;
}

File: leetcode/leetcode_python/course-schedule.py
class Solution:
    def canFinish_dfs_recursion(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        #build graph
        g = [[] for i in range(numCourses)]
        for i, j in prerequisites:
            g[j].append(i)
        
        v = [0] * numCourses
        def dfs(i):
            if v[i] == -1:
                return False
            if v[i] == 1:
                return True
            v[i] = -1
            for j in g[i]:
                if not dfs(j):
                    return False
            v[i] = 1
            return True
        
        for i in range(numCourses):
            if not dfs(i):
                return False
            
        return True
        
            
        
    def canFinish_bfs_interation(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # build graph
        g = [[] for i in range(numCourses)]
        indegree = [0] * numCourses
        for p in prerequisites:
            g[p[1]].append(p[0])
            indegree[p[0]] += 1
        
        bfs = []
        for i in range(numCourses):
            if (indegree[i] == 0):
                bfs.append(i)
                
        for i in bfs:
            for j in g[i]:
                indegree[j] -= 1
                if (indegree[j] == 0):
                    bfs.append(j)
                
        return len(bfs) == numCourses
        
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        #return self.canFinish_bfs_interation(numCourses, prerequisites)
        return self.canFinish_dfs_recursion(numCourses, prerequisites)
        